File: src/optical_element/MATLAB/Furion_cylinder_ellipse_Mirror.m
classdef Furion_cylinder_ellipse_Mirror<w_oe
    %波动光学圆柱镜  
    properties
        r; %圆柱镜面曲率半径  
        r1;
        r2;
        flag;
    end
    
    methods
        function obj=Furion_cylinder_ellipse_Mirror( beam_in,ds,di,chi,theta,surface,r1,r2,grating,flag)   
            obj = obj@w_oe( beam_in,ds,di,chi,theta,surface,grating) ;  %初始化父类  optical_element的参数      
           obj.r1=r1; %物距
            obj.r2 =r2;%像距
            obj.flag=flag;
            obj.reflect;%反射光束   
            
        end
    end   
    
      methods %面型反射,追迹
        function tracing(obj)
           obj.g_mirror = g_Furion_cylinder_ellipse_Mirror( obj.gbeam_in,obj.ds,obj.di,obj.chi,obj.theta,obj.surface,obj.r1,obj.r2,obj.grating)   ;
        end
      end
    
      methods   % 构造波动光束 不考虑畸变，将增加的相位直接加在原来的相位对应的位置之上；
        function  wbeam_out=creat_w_beam(obj,s_surface)     % 经过光学元器件后增加的相位（不考虑矩阵畸变）；
%         if obj.grating.m ==0          
%             wbeam=obj.beam_in;
%             Phase = obj.g_mirror.Phase + s_surface ;
%             phase1 = reshape(Phase,[wbeam.N, wbeam.N]);                             % 将增加的散点相位转化为二维网格矩阵相位      ；
%             Field_new = abs(wbeam.field).*exp(1i*(wbeam.phase_field+phase1));       % 通过增加的相位构建新的光的标量复数场分布    ；
% %             Field_new( isnan(Field_new))=0;
%             wbeam_out = beam(wbeam.X,wbeam.Y,Field_new,wbeam.wavelength);                 % 生成新的光束                              ；
%         else
            
             X_ =obj.g_mirror.X_;Y_ =obj.g_mirror.Y_;
            Phase =obj.g_mirror.Phase+s_surface;
            beam_in = obj.beam_in;
            phase_0 = Phase' + beam_in.phase_field(:);                        % 相位增量
           
                X=beam_in.XX;
                Y=beam_in.YY;      
                   
                    F = scatteredInterpolant(X_',Y_',phase_0);                    % 相位原坐标插值
                    F.Method = 'linear';
                    F.ExtrapolationMethod='linear';
                    phase_interp = F(X',Y');                            %用追迹后的坐标插值追迹前的相位
                    F1 = scatteredInterpolant(X_',Y_',abs(beam_in.field(:)));              %用追迹后的坐标插值追迹前的强度值
                    F1.Method = 'linear';
                    F1.ExtrapolationMethod='linear';
                    field_interp =   F1(X',Y');%delta_h = obj.surface.value(obj.a.Z2,obj.a.X2);
                    
          
            phase10 = reshape(phase_interp,[beam_in.N, beam_in.N]); % 一维数组转换为二维数组
            field10 = reshape(field_interp,[beam_in.N, beam_in.N]); % 一维数组转化为二维数组
            X_new1=reshape(X,[beam_in.N, beam_in.N]);
             Y_new1=reshape(Y,[beam_in.N, beam_in.N]);
             if (obj.flag ==0)
                 Field_new = abs(field10).*exp(1i*(phase10)); 
                  wbeam_out = beam(X_new1,Y_new1,Field_new,beam_in.wavelength);  %产生新的光束 
             else
            %
            L=max(max(beam_in.X))-min(min(beam_in.X));
              x = linspace(-L/2,L/2,10001);                                        %重新采样x轴方向
            y = x';                                                             %重新采样Y轴方向
            [X_r,Y_r] = meshgrid(x,y);   
             phase1 = interp2(X_new1,Y_new1,phase10,X_r,Y_r,'cubic',0); 
             field1 = interp2(X_new1,Y_new1,field10,X_r,Y_r,'cubic',0); 
            %
            Field_new = abs(field1).*exp(1i*(phase1)); %    通过插值后的振幅和相位组合成新的复数场 
%             X_new=reshape(X,[beam_in.N, beam_in.N]);
%             Y_new=reshape(Y,[beam_in.N, beam_in.N]);
            wbeam_out = beam(X_r,Y_r,Field_new,beam_in.wavelength);  %产生新的光束 
             end
        end
        end
%       end
end

File: src/optical_element/Furion_cylinder_ellipse_Mirror.cpp
#include "Furion_cylinder_ellipse_Mirror.h"

using namespace Furion_NS;

Furion_Cylinder_Ellipse_Mirror::Furion_Cylinder_Ellipse_Mirror(Beam* beam_in, double ds, double di, double chi, double theta, No_Surfe* surface, double r1, double r2, Grating* grating, double flag)
    : W_Oe(beam_in, ds, di, chi, theta, surface, grating), r1(r1), r2(r2), flag(flag)
{
    cout << "Furion_Cylinder_Ellipse_Mirror 初始化" << endl;
    //test();
}

Furion_Cylinder_Ellipse_Mirror::~Furion_Cylinder_Ellipse_Mirror()
{
    //delete ct;
    delete this->beam_in ;
}

void Furion_Cylinder_Ellipse_Mirror::run(Beam* beam_in, double ds, double di, double chi, double theta, No_Surfe* surface, double r1, double r2, Grating* grating, double flag)
{
    cout << "Furion_Cylinder_Ellipse_Mirror的run" << endl;
    reflect(beam_in, ds, di, chi, theta);

}

std::string Furion_Cylinder_Ellipse_Mirror::tracing()
{
    cout << "Furion_Cylinder_Ellipse_Mirror的tracing" << endl;
    
    this->g_FCE_mirror = new G_Furion_Cylinder_Ellipse_Mirror(this->gbeam_in, this->ds, this->di, this->chi, this->theta, surface, this->r1, this->r2, this->grating);
    this->g_FCE_mirror->run(this->gbeam_in, this->ds, this->di, this->chi, this->theta, surface, this->r1, this->r2, this->grating);
    return ("Furion_Cylinder_Ellipse_Mirror");
}

void Furion_Cylinder_Ellipse_Mirror::create_w_beam(std::vector<double>& s_phase)
{
    if (1)
    {
        //double* phase_0 = new double[this->N2];
        //double* phase_interp = new double[this->N2];
        std::vector<double> phase_0(this->N2);
        std::vector<double> phase_interp(this->N2);
        for (int i = 0; i < this->N; i++) {
            for (int j = 0; j < this->N; j++) {
                phase_0[i * this->N + j] = this->g_FCE_mirror->Phase[i * this->N + j] + s_phase[i * this->N + j] + beam_in->phase_field[i][j];
            }
        }
        Furion_NS::scatteredInterpolant(phase_interp, this->beam_in->XX, this->beam_in->YY, this->g_FCE_mirror->X_, this->g_FCE_mirror->Y_, phase_0, this->N);

        //double* field_interp = new double[this->N2];
        std::vector<double> field_interp(this->N2);
        for (int i = 0; i < this->N; i++){
            for (int j = 0; j < this->N; j++){
                phase_0[i * this->N + j] = std::abs(this->beam_in->field[i][j]);
            }
        }
        Furion_NS::scatteredInterpolant(field_interp, this->beam_in->XX, this->beam_in->YY, this->g_FCE_mirror->X_, this->g_FCE_mirror->Y_, phase_0, this->N);
        //destory_1d(phase_0);
        phase_0.clear();

        //double** X_new1;
        //double** Y_new1;
        //create_2d(X_new1, this->N);
        //create_2d(Y_new1, this->N);
        std::vector<std::vector<double> > X_new1(N, std::vector<double>(N));
        std::vector<std::vector<double> > Y_new1(N, std::vector<double>(N));

        Furion_NS::reshape(X_new1, this->beam_in->XX, this->N, this->N);
        Furion_NS::reshape(Y_new1, this->beam_in->YY, this->N, this->N);


        //std::complex<double>** Field_new;
        //create_2d(Field_new, this->N);
        std::vector<std::vector<std::complex<double>> > Field_new(N, std::vector<std::complex<double>>(N));
        if (this->flag == 0)
        {
            for (int i = 0; i < this->N; i++) {
                for (int j = 0; j < this->N; j++) {
                    Field_new[i][j] = std::fabs(field_interp[i * N + j]) * std::exp(std::complex<double>(0, phase_interp[i * N + j]));
                }
            }
            //destory_1d(phase_interp);
            //destory_1d(field_interp);
            phase_interp.clear();
            field_interp.clear();

            this->beam_out = new Beam(X_new1, Y_new1, Field_new, this->beam_in->wavelength, N);

        }
        else
        {
            double L = Furion_NS::Max(this->beam_in->X[0], N) - Furion_NS::Min(this->beam_in->X[0], N);
            //double* x = new double[this->N];
            //double* y = new double[this->N];
            //double* y_new1 = new double[this->N];
            std::vector<double> x(this->N);
            std::vector<double> y(this->N);
            std::vector<double> y_new1(this->N);

            
            //double* phase1 = new double[this->N2];
            //double* field1 = new double[this->N2];
            std::vector<double> phase1(this->N2);
            std::vector<double> field1(this->N2);

            Furion_NS::linspace(x, -L / 2, L / 2, this->N);
            for (int i = 0; i < this->N; i++) {
                y[i] = x[i];
                y_new1[i] = Y_new1[i][0];
                //cout << y_new1[i] << endl;
            }

            Furion_NS::interp2_1(phase1, X_new1[0], y_new1, phase_interp, x, y, this->N2, this->N, this->N, "cubic");
            //exit(0);
            Furion_NS::interp2_1(field1, X_new1[0], y_new1, field_interp, x, y, this->N2, this->N, this->N, "cubic");
            //destory_1d(x);
            //destory_1d(y);
            //destory_1d(y_new1);
            //destory_1d(phase_interp);
            //destory_1d(field_interp);
            x.clear();
            y.clear();
            y_new1.clear();
            phase_interp.clear();
            field_interp.clear();

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    Field_new[i][j] = std::fabs(phase1[i * N + j]) * std::exp(std::complex<double>(0, field1[i * N + j]));
                }
            }
            //destory_1d(phase1);
            //destory_1d(field1);
            phase1.clear();
            field1.clear();

            this->beam_out = new Beam(X_new1, Y_new1, Field_new, this->beam_in->wavelength, N);
        }
    }
}






